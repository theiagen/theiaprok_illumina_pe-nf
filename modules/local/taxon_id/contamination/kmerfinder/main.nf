process KMERFINDER_BACTERIA {
    tag "$meta.id"
    label 'process_high_memory'

    container "us-docker.pkg.dev/general-theiagen/biocontainers/kmerfinder:3.0.2--hdfd78af_0"

    input:
    tuple val(meta), path(assembly)
    path kmerfinder_db

    output:
    tuple val(meta), path("*_kmerfinder.tsv"), emit: kmerfinder_results_tsv
    tuple val(meta), path("TOP_HIT"),           emit: kmerfinder_top_hit
    tuple val(meta), path("QC_METRIC"),         emit: kmerfinder_query_coverage
    tuple val(meta), path("TEMPLATE_COVERAGE"), emit: kmerfinder_template_coverage
    tuple val(meta), path("DATABASE"),          emit: kmerfinder_database
    path("versions.yml"), emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ""
    
    """
    # Decompress the kmerfinder bacterial database
    mkdir db
    tar -C ./db/ -xzvf ${kmerfinder_db}  
    ls ./db
    
    # Run kmerfinder
    kmerfinder.py \\
        -db ./db/bacteria/bacteria.ATG \\
        -tax ./db/bacteria/bacteria.tax \\
        -i ${assembly} \\
        -o ${prefix} \\
        ${args} 

    # parse outputs
    if [ ! -f ${prefix}/results.txt ]; then
      PF="No hit detected in database"
      QC="No hit detected in database"
      TC="No hit detected in database"
    else
      PF="\$(cat ${prefix}/results.txt | head -n 2 | tail -n 1 | cut -f 19)"
      QC="\$(cat ${prefix}/results.txt | head -n 2 | tail -n 1 | cut -f 6)"
      TC="\$(cat ${prefix}/results.txt | head -n 2 | tail -n 1 | cut -f 7)"
        # String is empty or just contains the header
        if [ "\$PF" == "" ] || [ "\$PF" == "Species" ]; then
          PF="No hit detected in database"
          QC="No hit detected in database"
          TC="No hit detected in database"
        fi
      mv -v ${prefix}/results.txt ${prefix}_kmerfinder.tsv
    fi
    echo \$PF | tee TOP_HIT
    echo \$QC | tee QC_METRIC
    echo \$TC | tee TEMPLATE_COVERAGE

    # extract database name
    DB=\$(basename ${kmerfinder_db} | sed 's/\\.tar\\.gz\$//')
    echo \$DB | tee DATABASE

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kmerfinder: kmerfinder:3.0.2--hdfd78af_0
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "na" > "${prefix}_kmerfinder.tsv"
    echo "na" > "TOP_HIT"
    echo "na" > "QC_METRIC"
    echo "na" > "TEMPLATE_COVERAGE"
    echo "na" > "DATABASE"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kmerfinder: kmerfinder:3.0.2--hdfd78af_0
    END_VERSIONS
    """
}