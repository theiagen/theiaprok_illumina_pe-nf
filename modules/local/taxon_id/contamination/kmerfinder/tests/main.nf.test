nextflow_process {
    name "Test Process KMERFINDER_BACTERIA"
    script "../main.nf"
    process "KMERFINDER_BACTERIA"
    
    test("kmerfinder test") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz", checkIfExists: true)
                ])
                input[1] = file("$projectDir/../../datasets/kmerfinder_bacteria_20230911.tar.gz", checkIfExists: true)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.kmerfinder_results_tsv,
                    process.out.kmerfinder_top_hit,
                    process.out.kmerfinder_query_coverage,
                    process.out.kmerfinder_template_coverage,
                    process.out.kmerfinder_database,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("kmerfinder - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz", checkIfExists: true)
                ])
                input[1] = file("$projectDir/../../datasets/kmerfinder_bacteria_20230911.tar.gz", checkIfExists: true)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}