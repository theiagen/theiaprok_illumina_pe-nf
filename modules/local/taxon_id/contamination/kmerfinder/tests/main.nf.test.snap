{
    "kmerfinder test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_kmerfinder.tsv:md5,1580af2fae93812869f04708639f5314"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "TOP_HIT:md5,cf60bcbd06725055f50468a1f26bab9f"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "QC_METRIC:md5,17c62290ed5713b447492e2730a41622"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "TEMPLATE_COVERAGE:md5,5acd6ae3f246733a2e8e3ad364218579"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "DATABASE:md5,ff5cb9ab6b0e006ce178e5f89cbf4ea7"
                ]
            ],
            [
                "versions.yml:md5,492737a83d2c59c1357ff111997036e6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-24T14:54:19.187995136"
    },
    "kmerfinder - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_kmerfinder.tsv:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TOP_HIT:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "QC_METRIC:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TEMPLATE_COVERAGE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "DATABASE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "5": [
                    "versions.yml:md5,492737a83d2c59c1357ff111997036e6"
                ],
                "kmerfinder_database": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "DATABASE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "kmerfinder_query_coverage": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "QC_METRIC:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "kmerfinder_results_tsv": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_kmerfinder.tsv:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "kmerfinder_template_coverage": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TEMPLATE_COVERAGE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "kmerfinder_top_hit": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TOP_HIT:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,492737a83d2c59c1357ff111997036e6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-24T14:54:24.368894277"
    }
}